oidc-exchange.py

```diff
@@ -0,0 +1,161 @@
+def die(msg: str) -> NoReturn:
+    with _GITHUB_STEP_SUMMARY.open("a", encoding="utf-8") as io:
+        print(msg, file=io)
+
+    # NOTE: `msg` is Markdown formatted, so we emit only the header line to
+    # avoid clogging the console log with a full Markdown formatted document.
+    header = msg.splitlines()[0]
+    print(f"::error::OIDC exchange failure: {header}", file=sys.stderr)
+    sys.exit(1)
+
+
+def debug(msg: str):
+    print(f"::debug::{msg}", file=sys.stderr)
+
+
+def get_normalized_input(name: str) -> str | None:
+    name = f"INPUT_{name.upper()}"
+    if val := os.getenv(name):
+        return val
+    return os.getenv(name.replace("-", "_"))
+
+
+def assert_successful_audience_call(resp: requests.Response, domain: str):
+    if resp.ok:
+        return
+
+    match resp.status_code:
+        case 403:
+            # This index supports OIDC, but forbids the client from using
+            # it (either because it's disabled, ratelimited, etc.)
+            die(f"audience retrieval failed: repository at {domain} has OIDC disabled")
+        case 404:
+            # This index does not support OIDC.
+            die(
+                "audience retrieval failed: repository at "
+                f"{domain} does not indicate OIDC support",
+            )
+        case other:
+            # Unknown: the index may or may not support OIDC, but didn't respond with
+            # something we expect. This can happen if the index is broken, in maintenance mode,
+            # misconfigured, etc.
+            die(
+                "audience retrieval failed: repository at "
+                f"{domain} responded with unexpected {other}",
+            )
+
+
+repository_url = get_normalized_input("repository-url")
+is_pypi = not repository_url
+if is_pypi:
+    # Easy case: no explicit repository URL, which means we're using PyPI and we can just
+    # hardcode the exchange endpoint and OIDC audience.
+    token_exchange_url = "https://pypi.org/_/oidc/github/mint-token"
+    oidc_audience = "pypi"
+else:
+    repository_domain = urlparse(repository_url).netloc
+    token_exchange_url = f"https://{repository_domain}/_/oidc/github/mint-token"
+
+    # Indices are expected to support `https://{domain}/_/oidc/audience`,
+    # which tells OIDC exchange clients which audience to use.
+    audience_url = f"https://{repository_domain}/_/oidc/audience"
+    audience_resp = requests.get(audience_url)
+    assert_successful_audience_call(audience_resp, repository_domain)
+
+    oidc_audience = audience_resp.json()["audience"]
+
+debug(f"selected OIDC token exchange endpoint: {token_exchange_url}")
+
+try:
+    oidc_token = id.detect_credential(audience=oidc_audience)
+except id.IdentityError as identity_error:
+    die(_TOKEN_RETRIEVAL_FAILED_MESSAGE.format(identity_error=identity_error))
+
+# Now we can do the actual token exchange.
+mint_token_resp = requests.post(
+    token_exchange_url,
+    json={"token": oidc_token},
+)
+
+try:  # noqa: WPS229
```

Comment:
This doesn't seem necessary.
```suggestion
try:
```
If the linting violation is still happening, it should be fixed, not ignored:
> Forbid try blocks with bodies that are too long.